tensorly/backend/core.py

```diff
@@ -0,0 +1,854 @@
+
+        Parameters
+        ----------
+        matrix : tensor
+            A 2D tensor.
+        n_eigenvecs : int, optional, default is None
+            If specified, number of eigen[vectors-values] to return.
+
+        Returns
+        -------
+        U : 2-D tensor, shape (matrix.shape[0], n_eigenvecs)
+            Contains the right singular vectors
+        S : 1-D tensor, shape (n_eigenvecs, )
+            Contains the singular values of `matrix`
+        V : 2-D tensor, shape (n_eigenvecs, matrix.shape[1])
+            Contains the left singular vectors
+        """
+        # Check that matrix is... a matrix!
+        if self.ndim(matrix) != 2:
+            raise ValueError('matrix be a matrix. matrix.ndim is %d != 2'
+                             % self.ndim(matrix))
+
+        ctx = self.context(matrix)
+        is_numpy = isinstance(matrix, np.ndarray)
+
+        if not is_numpy:
+            matrix = self.to_numpy(matrix)
+
+        # Choose what to do depending on the params
+        dim_1, dim_2 = matrix.shape
+        if dim_1 <= dim_2:
+            min_dim = dim_1
+            max_dim = dim_2
+        else:
+            min_dim = dim_2
+            max_dim = dim_1
+
+        if n_eigenvecs >= min_dim:
+            if n_eigenvecs > max_dim:
+                warnings.warn(('Trying to compute SVD with n_eigenvecs={1}, which '
+                               'is larger than max(matrix.shape)={1}. Setting '
+                               'n_eigenvecs to {1}').format(n_eigenvecs, max_dim))
+                n_eigenvecs = max_dim
+
+            if n_eigenvecs is None or n_eigenvecs > min_dim:
+                full_matrices = True
+            else:
+                full_matrices = False
+
+            # Default on standard SVD
+            U, S, V = scipy.linalg.svd(matrix, full_matrices=full_matrices)
+            U, S, V = U[:, :n_eigenvecs], S[:n_eigenvecs], V[:n_eigenvecs, :]
+        else:
+            # We can perform a partial SVD
+            # First choose whether to use X * X.T or X.T *X
+            if dim_1 < dim_2:
+                S, U = scipy.sparse.linalg.eigsh(
+                    np.dot(matrix, matrix.T.conj()), k=n_eigenvecs, which='LM'
+                )
+                S = np.sqrt(S)
+                V = np.dot(matrix.T.conj(), U * 1 / S[None, :])
+            else:
+                S, V = scipy.sparse.linalg.eigsh(
+                    np.dot(matrix.T.conj(), matrix), k=n_eigenvecs, which='LM'
+                )
+                S = np.sqrt(S)
+                U = np.dot(matrix, V) * 1 / S[None, :]
+
+            # WARNING: here, V is still the transpose of what it should be
+            U, S, V = U[:, ::-1], S[::-1], V[:, ::-1]
+            V = V.T.conj()
+
+        if not is_numpy:
+            U = self.tensor(U, **ctx)
+            S = self.tensor(S, **ctx)
+            V = self.tensor(V, **ctx)
+        return U, S, V
+
+
+def dispatch(method):
```

Comment:
Is the purpose of this function to dynamically update the backend being used even in the case where it was imported as 
` from tensorly.backend import some_fun` ? 

Is there a "cleaner" way to do this?

Reply:
> Is the purpose of this function to dynamically update the backend being used even in the case where it was imported as
> `from tensorly.backend import some_fun` ?

Yes. The purpose of this is:
- Dispatch methods to the current backend at call time
- Provide stable functions and documentation on all top-level methods (i.e. `tl.ones` is always the same object, regardless of backend)

> Is there a "cleaner" way to do this?

I don't think this is particularly magical, and can't see a simpler way. We're creating methods to call the specific backend functions. If a new method is added, the author needs to:

- Add it to the `Backend` class, with a docstring
- Implement it in whatever backends are needed
- Add a line here creating a new dispatched function

These are process steps, can be documented (happy to add this in a subsequent PR), and don't seem particularly tricky. Methods removing any of these steps would be more magical and [less explicit](https://www.python.org/dev/peps/pep-0020/#id3) (IMHO).

Reply:
The dispatch function is pretty neat. My issue is more with having this dynamical dispatch at call time: I would be more in favour of dynamically updating the imported functions when needed.

Back to the previous comment, the dispatching could be done by the BackendManager.. 

Clear, explicit, readable code **is** a guiding principle in TensorLy ;)